File: src/main/java/algo/backtrack/PartitionToKEqualSumSubsets_698.java
package algo.backtrack;

import java.util.HashMap;

public class PartitionToKEqualSumSubsets_698 {
  public static boolean canPartitionKSubsets(int[] nums, int k) {
    int len = nums.length, sum = 0;
    var count = new HashMap<Integer, Integer>();
    for (var n : nums) {
      sum += n;
      count.put(n, count.getOrDefault(n, 0) + 1);
    }
    if (sum % k != 0) {
      return false;
    }

    var partitionSum = sum / k;
    return backtrack(nums, k, partitionSum, partitionSum, 0, new int[k + 1][1 << (len + 1)]);
  }

  private static boolean backtrack(int[] nums, int k, int psum, int left, int bitmask, int[][] dp) {
    if (left == 0 && k == 1) {
      return true;
    }
    // bitmask is the binary representation of if nums[i] is used
    if (dp[k][bitmask] != 0) {
      return dp[k][bitmask] == 1;
    }
    if (left == 0) {
      k--;
      left = psum;
    }
    var len = nums.length;
    for (var i = 0; i < len; i++) {
      if (nums[i] > psum) {
        return false;
      }
      var pos = 1 << (len - i);
      if ((bitmask & pos) != 0) {
        continue;
      }
      if (left >= nums[i]) {
        var mask = bitmask ^ pos;
        if (backtrack(nums, k, psum, left - nums[i], mask, dp)) {
          dp[k][mask] = 1;
          return true;
        }
        dp[k][mask] = -1;
      }
    }
    return false;
  }

  public static void main(String[] args) {
//    System.out.println(canPartitionKSubsets(new int[]{4,3,2,3,5,2,1}, 4));
//    System.out.println(canPartitionKSubsets(new int[]{1,2,3,4}, 3));
//    System.out.println(canPartitionKSubsets(new int[]{129,17,74,57,1421,99,92,285,1276,218,1588,215,369,117,153,22}, 3));
//    System.out.println(canPartitionKSubsets(new int[]{15,3557,42,3496,5,81,34,95,9,81,42,106,71}, 11));
//    System.out.println(canPartitionKSubsets(new int[]{3,3,10,2,6,5,10,6,8,3,2,1,6,10,7,2}, 6));
//    System.out.println(canPartitionKSubsets(new int[]{99,37,37,37,37,37,37,37,37,5}, 4));
    System.out.println(canPartitionKSubsets(new int[]{5,5,5,5,4,4,4,4,3,3,3,3}, 4));
  }
}

File: src/main/kotlin/algo/backtrack/PartitionToKEqualSumSubsets_698.kt
package algo.backtrack

class KtPartitionToKEqualSumSubsets_698 {
    fun canPartitionKSubsets(nums: IntArray, k: Int): Boolean {
        val total = nums.sum()
        if (total % k != 0) {
            return false
        }
        val target = total / k

        fun backtrack(nums: IntArray, sum: Int, subsetNum: Int, bitmask: Int, dp: Array<IntArray>): Int {
            if (sum == total && subsetNum == k) {
                return 1
            }
            if (dp[subsetNum][bitmask] != 0) {
                return dp[subsetNum][bitmask]
            }
            val currentSum = sum % target
            for (i in nums.indices) {
                val pos = 1 shl (nums.size - i)
                if (pos and bitmask != 0) {
                    continue
                }
                val nextSum = currentSum + nums[i]
                if (nextSum > target) {
                    continue
                }
                val nextSubsetNum = if (nextSum == target) subsetNum + 1 else subsetNum
                val res = backtrack(nums, sum + nums[i], nextSubsetNum, pos xor bitmask, dp)
                dp[subsetNum][bitmask] = res
                if (res == 1) {
                    break
                }
                dp[subsetNum][bitmask] = -1
            }
            return dp[subsetNum][bitmask]
        }

        val dp = Array(k + 1) { IntArray(1 shl (nums.size + 1)) {0} }
        return backtrack(nums, 0, 0, 0, dp) == 1
    }
}

fun main() {
    val kpartition = KtPartitionToKEqualSumSubsets_698()
    println(kpartition.canPartitionKSubsets(intArrayOf(4,3,2,3,5,2,1), 4))
    println(kpartition.canPartitionKSubsets(intArrayOf(1,2,3,4), 3))
}